### ISSUE TITLE
gtk version panel icon click displays the context menu only

### ISSUE THREAD

**Issue by ineuw** [autokey/autokey#454]
## Classification:
UI/Usability

## Reproducibility:
Always

## Version
95.10 and earlier GTK versions

AutoKey version: 
95.10-gtk

If the problem is known to be present in more than one version, please list all of those.
95.10 and earlier GTK versions This never worked, not even in the 2.7 versions

Linux Distribution:
Installed from Linux Mint Repository

Summary of the problem:
Clicking on the panel icon with either left or right mouse button, it displays only the context menu. The left mouse button should activate the main window. To see the main window is an additional step in the Gtk version.

The main window appears only once when Autokey is loaded at startup.

- I do this
Mouse left click on the panel icon

## Expected Results
To see main window

- Instead, this happens. :(
I see the context menu

**Comment by josephj11** [comment]
Verified. This is an autokey-gtk problem. autokey-qt works correctly.

**Comment by sebastiansam55** [comment]
https://lazka.github.io/pgi-docs/AppIndicator3-0.1/classes/Indicator.html#AppIndicator3.Indicator.set_menu
https://askubuntu.com/questions/930848/how-to-program-a-one-click-indicator-add-middle-click-functionality

It doesn't seem like gtk is meant to work this way? Has it ever worked like this?

The only option for AppIndicator3 (which AutoKey uses) for primary click is `set_menu` which only takes a `Gtk.Menu` object, which is the popup we currently see when click on the indicator. 
There is a option for setting what happens when you click with scrollwheel button but that takes a Gtk.Widget object which I'm not sure how to implement something to do to bring up the window with that.

EDIT:
You can use autokey script `system.exec_command('autokey-gtk')`, and bind it to whatever you want. 
Which while it will call the "system" install version of AutoKey it will open the window for the running instance of autokey-gtk.

**Comment by josephj11** [comment]
Thanks for looking. IDK if it ever worked. I have been using the Qt version for a long time. Long ago (original branch), I used the GTK version because it was usually ahead of the Qt version, but that's too long ago to remember specifics.

**Comment by BlueDrink9** [comment]
Regarding the tray icon for GTK, [I found this explanation](https://github.com/electron/electron/issues/21576#issuecomment-580421902) about the process. We are using AppIndicator for the GTK tray menu for whatever reason, and the standard behavior for that should be left click to view items. I'm not sure if it can be changed.


[This wiki page](https://wiki.ubuntu.com/DesktopExperienceTeam/ApplicationIndicators#Porting_Guide_for_Applications) suggests the API should be fairly similar to [GtkStatusIcon](https://developer.gnome.org/gtk3/stable/GtkStatusIcon.html). I'm not seeing anything there about clicking, only a single "activation". [Possible hope in this comment but probably a bit of a stretch](https://github.com/transmission/transmission/issues/1387#issuecomment-793012606) especially given the response to [this bug report](https://bugs.launchpad.net/libappindicator/+bug/1428473)

**Comment by ineuw** [comment]
@BlueDrink9, thanks for the links. I was very curious about it, but not at the price of your precious time. I am sure you have more important things to do. I will read up and try to collect some more info if I understand it. I would close the issue because of it's low priority.

**Comment by BlueDrink9** [comment]
Quite alright, I was curious and I thought it would be a quick fix. It was something that mildly bugged me as well

**Comment by BlueDrink9** [comment]
Still an issue, and a legitimate one. Just because we can't fix it right now, doesn't mean it's over and finished with so we'll leave it open for the time being.

**Comment by josephj11** [comment]
As a workaround, it would be very easy to write a bash script that runs `autokey-gtk -c`, give it an icon, and place that icon in your panel or on your desktop. I do things like this all the time in KDE, but it is probably easy enough to do in whatever desktop environment you are using. I'd be happy to help figure it out if you are interested.

When you run autokey-gtk -c, it will display the AutoKey main menu. If AutoKey is already running, that's all it will do. If it's not running, it will launch it and then display the main menu.

**Comment by ineuw** [comment]
Assigned Ctrl+1 to activate the main window, but a panel icon has become very desirable. Having quick access to the Main panel 
with either methods is important  because of changes I must make .  . . . .(all was needed is the menu context option to add it to the panel and it's done). Thanks

**Comment by josephj11** [comment]
I don't quite understand your last comment. So, you have your own panel icon now?

Why did you close this issue? It may not be a high priority to fix this, but it still appears broken to me.

**Comment by ineuw** [comment]
I closed it because it's really not high priority. But after working for the past few days with two substitute methods, I agree with you 100% that it's inconvenient to work without the native tray menu. 

I tried both substitutes. Opening the main menu with a keyboard shortcut, or substituting the native tray icon with a panel shortcut, are inconvenient because neither indicate the true status of the app, and makes the quitting process cumbersome.

**Comment by josephj11** [comment]
Clarification: my workaround was to add a new icon to the panel, not to replace the original one.

Also, a bug/deficiency is still a problem even if fixing it is a low priority. It's good to have it noted because when someone is fixing something else in the affected code, they may see a way to fix it or they may be considering doing something else which might cause a similar issue and see that it might not be a good idea...

You can certainly indicate how critical you feel an issue is when you file it. We don't have a severity heading in our issues template. We could easily add one if people think it would help.

**Comment by ineuw** [comment]
I am re-activating this to find a way to correct this minor but irritating issue. I am using Linux Mint Cinnamon with v96.0 GTK, but the problem remains the same as it was as with version 95.xx GTK which is Python 2.

Linux Mint Cinnamon repaired the duplicate icon issue caused by their X-apps status app, but the panel menu activation with the mouse buttons was not repaired. I am getting the same menu.

Could this be a library reference issue in AK code, because of changes in Python? I could use help in tracking it down, because it AFAIK it works properly in the QT version.

The upper left corner of the photo shows the image in the 2020 post and the current photo in the lower right.


![221231_autokey_panel_issue](https://user-images.githubusercontent.com/10367016/210131787-10b7023c-3d85-423f-8071-28f5b4d5cf22.jpg)

**Comment by josephj11** [comment]
@ineuw First, this branch of AutoKey has always been Python3. 0.90.x was the last Python2 version.

The issue here is that the Qt and GTK front ends are based on completely separate API libraries and the design of the GTK library differs from that of Qt in this particular and doesn't appear to offer the option we desire.

Since this may be a capability many people/applications want, you might have some success if you can find a way to communicate with UI/UX designers/programmers in the GTK ecosystem and see if they have any workarounds for this issue or if there is an alternative to the AppIndicator code we use for interacting with our panel icon.

----

I am unclear about the workarounds you have tried and why they are not sufficient for your needs. Adding another icon to your panel is no big deal unless panel space is scarce and using a separate key binding is just a matter of repetition/muscle memory for it to become second nature.

**Comment by ineuw** [comment]
@josephj11, you misread my post regarding the Python version. All it meant that the problem persisted in both Autokey versions of both 95.xx Python2, and 96.0Python3.

Moved my language comment to where it belongs in #754

**Comment by ineuw** [comment]
Replying to the first part of your comment. I forgot to make it clear that I am only interested in the GTK version, because I can't swim in so many oceans. 

Much thanks for your reply which contains the needed info take for my next step.

**Comment by josephj11** [comment]
Just a couple of quick nit picking items:

What is "95.xx Python2"? That makes no sense to me.

I wasn't suggesting that you switch between our two front ends. I was noting that the panel icon works as desired in the  Qt front end and that that's because of the Qt API library behavior, not from something special we did in AutoKey and that the GTK API does not appear to provide similar functionality, so it's not something where we can just change a setting and get the desired behavior.

Good luck searching for another solution. We definitely could use it.

**Comment by ineuw** [comment]
I mentioned "95.xx Python2" to indicate that the panel menu problem existed in Autokey 95.xx with Python2. Then, I had no clue where the issue originated. I didn't know about GTK and assumed that the problem was Gnome. I downloaded the 96.0 code out of curiosity,

**Comment by BlueDrink9** [comment]
Yeah, sadly doesn't seem like GTK has any interest in changing this behaviour at the moment, so I think this has to remain a wontfix

**Comment by ineuw** [comment]
I found another related defect, Autokey "Show Main menu" doesn't abide by system windows setting, while all other apps do. This will help a lot. I will report whatever I am told by Linux Mint Cinnamon.

**Comment by josephj11** [comment]
> I mentioned "95.xx Python2" to indicate that the panel menu problem existed in Autokey 95.xx with Python2. Then, I had no clue where the issue originated. I didn't know about GTK and assumed that the problem was Gnome. I downloaded the 96.0 code out of curiosity,

I know this is a small thing, but there is no such thing as "95.xx Python2" . This branch of AutoKey was created with the express purpose of converting AutoKey from Python 2 to Python 3. I think it started with 0.93, but I don't remember the exact release number. The last version of AutoKey that used Python 2 was 0.90.x, around eight years ago.

**Comment by josephj11** [comment]
> I found another related defect, Autokey "Show Main menu" doesn't abide by system windows setting, while all other apps do. This will help a lot. I will report whatever I am told by Linux Mint Cinnamon.

What exactly are you referring to here? The window theme? How can I see/recreate this on my system?

**Comment by ineuw** [comment]
https://github.com/linuxmint/cinnamon/issues/11392

**Comment by josephj11** [comment]
@bluedrink9 Can you translate the above link into English? I sort of get it, but it looks like there is an ancestor of AppIndicator that might work correctly for us - at the cost of taking us a bit further from Wayland compatibility.

@ineuw - that answers my previous question.

**Comment by josephj11** [comment]
@ineuw Who closed the LMC issue above? I don't seem to see that.

**Comment by BlueDrink9** [comment]
> @bluedrink9 Can you translate the above link into English? I sort of get it, but it looks like there is an ancestor of AppIndicator that might work correctly for us - at the cost of taking us a bit further from Wayland compatibility.
> 
> @ineuw - that answers my previous question.

My reading isn't that it's an ancestor, but current possibility. Personally not going to do it though

**Comment by ineuw** [comment]
@BlueDrink9 @josephj11 my sincere apologies for the late reply. I crashed and reinstalled everything anew. 

Thanks for looking at the post on Cinnamon. If nothing can be done about it, it's understood. I also think, that is why they closed it. Also, X11 is a weakness for Autokey.
